### PR TITLE
JP-3357: Pathloss: Implement NIRSPEC MSA correction for slits that aren't 1x1 or 1x3

### LINKED ISSUE #7864: Pathloss corrections for long MOS slitlets
_Issue [JP-3357](https://jira.stsci.edu/browse/JP-3357) was created on JIRA by [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke):_

This ticket is associated with a known issue in NIRSpec MOS data products: spectra extracted from slitlets consisting of more than 3 shutters exhibit unexpected wavelength-dependent flux discrepancies.

The issue appears to be that the pathloss correction file contains data with explicit corrections for 1 shutter slitlets and 3 shutter slitlets.  If the input data matches one of these two configurations, the correction is applied.  If not, no correction is applied and the pipeline issues a warning like: "WARNING: Cannot find matching pathloss model for slit with5 shutters".

To properly correct arbitrary MSA configurations, the pipeline needs to implement a strategy to compose a correction from the existing reference data.    In particular, the correction derived for a 3-shutter slitlet should be applicable to any slitlet > 1 shutter in length because it considers light from a source that may extend only into adjacent shutters. So, for a source in any of the middle 3 shutters of a 5-shutter slitlet, for example, the pipeline should apply a correction based on the middle shutter of the 3-shutter case in the reference file. If the source is in the top or bottom shutter, then it should use the corresponding top/bottom shutter in the reference file.

### PR DESCRIPTION
**Pull request by stscirij**
Resolves [JP-3357](https://jira.stsci.edu/browse/JP-3357)

Closes #7864

This PR fixes the deficiency of the pathloss correction that was not applied if the science data slit is not 1x1 or 1x3.
See JP-3357 and the updated docs for details
 
**Checklist for maintainers**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [x] updated or added relevant tests
- [x] updated relevant documentation
- [x] added relevant milestone
- [x] added relevant label(s)
- [x] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

File: jwst/pathloss/pathloss.py [hunk 1]
```diff
@@ -237,6 +261,60 @@ def calculate_pathloss_vector(pathloss_refdata,
         return wavelength, pathloss_vector, is_inside_slitlet
 
 
+def calculate_two_shutter_uniform_pathloss(pathloss_model):
+    """The two shutter MOS case for uniform source calculation requires a custom
+     routine since it uses both the 1X1 and 1X3 extensions of the pathloss reference file
+
+    Parameters
+     ---------
```
Comment (hbushouse) [inline jwst/pathloss/pathloss.py:269]:
```suggestion
    ----------
```

**Review by hbushouse — COMMENTED** [review]
Just one minor nit.

**Comment by hbushouse** [comment]
@stscirij I can't get the regtests to run on this PR branch. Some kind of problem with the branch being based on a very old tag from master. Need to update the PR branch to include all tags. And then please start a regtest run over at https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/

**Comment by stscirij** [comment]
Ran a regtest run at
[https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1081/](https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1081/)

File: docs/jwst/pathloss/description.rst [hunk 2]
```diff
@@ -49,6 +49,58 @@ The form of the 2-D correction (point or uniform) that's appropriate for the
...
+ array, since each pixel has a different wavelength, and the correction is applied
+ to the science pixel array.
+
+ For uniform sources, there is no dependence of the pathloss correction on position,
+ so the correction arrays are just 1-d arrays of correction and wavelength.  Once
+ again, the shutter_state attribute of each slit is used to determine the correction
+ entry used:
+
+ * if both shutters adjacent to the fiducial are closed, the 1x1 entry is used
+ * if both shutters adjecent to the fiducial are open, the 1x3 entry is used
```
Comment (hbushouse) [inline docs/jwst/pathloss/description.rst:96]:
```suggestion
 * if both shutters adjacent to the fiducial are open, the 1x3 entry is used
```

File: docs/jwst/pathloss/description.rst [hunk 3]
```diff
@@ -49,6 +49,58 @@ The form of the 2-D correction (point or uniform) that's appropriate for the
...
+ * if one is closed and 1 is open, the correction used is the average of the 1x1
```
Comment (hbushouse) [inline docs/jwst/pathloss/description.rst:97]:
```suggestion
 * if one is closed and one is open, the correction used is the average of the 1x1
```

**Comment by nden** [comment]
> I deleted the tag from my fork to get it to run

What you should do to avoid problems in the future is push all tags to your fork
```
git fetch upstream
git push origin --tags
```

File: docs/jwst/pathloss/description.rst (hunk 2 above)
Comment (stscirij) [reply docs/jwst/pathloss/description.rst]: Fixed

File: docs/jwst/pathloss/description.rst (hunk 3 above)
Comment (stscirij) [reply docs/jwst/pathloss/description.rst]: Fixed

**Review by hbushouse — APPROVED** [review]
Looks good.

File: docs/jwst/pathloss/description.rst [hunk 4]
```diff
@@ -49,6 +49,58 @@ The form of the 2-D correction (point or uniform) that's appropriate for the
...
+of the source (ranging from -0.5 to 0.5 in both X and Y), while the third dimension
+carries the wavelength dependence.  The 1x3 shutter is 3 times as large in Y as in X.
+
+The entry to use for a point source target is determined by looking at the shutter_state
+attribute of the slit used.  This is a string with a length equal to the number
+of shutters that make up the slit, with 1 denoting an open shutter, 0 a closed
+shutter and x the fiducial (target) shutter.  The reference entry is determined
+by how many shutters next to the fiducial shutter are open:
+
+ * if both adjacent shutters are closed, the 1x1 entry is used.  A matching
```
Comment (hbushouse) [inline docs/jwst/pathloss/description.rst:66]: aaaargh. Just looked at the resulting rtd pages and the formatting of this whole section that's supposed to use bullets (or whatever) is wonky. Can you try some kind of different formatting?

Comment (hbushouse) [reply docs/jwst/pathloss/description.rst:66]: See https://jwst-pipeline--8106.org.readthedocs.build/en/8106/jwst/pathloss/description.html#nirspec

**Pull request merged**